### PR TITLE
Native OpenMP benchmark

### PR DESCRIPTION
**Pull request by csegarragonz**
In this PR I bring the CovidSim experiment up to speed and prepare the native version of the benchmark. Results can be seen for simulations in [Malta](https://github.com/faasm/experiment-base/blob/0dc848ff4ff11ecc4f8d58cfeaa23f7bd0b48b6a/results/covid/covid_malta.png) and [Guam](https://github.com/faasm/experiment-base/blob/0dc848ff4ff11ecc4f8d58cfeaa23f7bd0b48b6a/results/covid/covid_guam.png). The graphs present the elapsed time to run (setup time subtracted) a _no intervention_ model in the country with a certain number of OpenMP threads, for which I report the average and standard deviation.

Most notably, in this PR I build a docker image tagged `faasm/experiment-covid-native` with the necessary dependencies and software to run the simulation installed. It can be run either:
* LIke a standalone script: `docker run --rm faasm/experiment-covid-native:0.0.1 <Country> <Num_threads>`
* Or bashed into by overriding the entry point: `docker run --rm -it --entrypoint bash faasm/experiment-covid-native:0.0.1`

Building and running should be straightforward and run without problems. The only changes upstream (i.e. in [faasm/covid-sim](https://github.com/faasm/covid-sim)) are related to timing issues discussed offline.

Lastly, running `run/native.sh` will also create a plot in `experiment-base/results/covid/covid.png`. Currently the experiment parameters are held in `./run/all_native.py` but they could (should) be moved elsewhere eventually.

### CODE REVIEW

File: bin/build_native.sh [hunk 1]
```diff
@@ -2,25 +2,36 @@
```
Comment (csegarragonz) [inline bin/build_native.sh:2]: This script builds a container tagged `faasm/experiment-covid-native:VERSION` with the necessary dependencies and software to run `CovidSim`.

File: docker/experiment-covid-native.dockerfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+WORKDIR /code
+RUN git clone https://github.com/faasm/experiment-covid
+
+# Build native 
+WORKDIR /code/experiment-covid/
+RUN git submodule update --init
+RUN ./build/native.sh
+
+# Unzip and copy population files
```
Comment (csegarragonz) [inline docker/experiment-covid-native.dockerfile:15]: This is to allow us running experiments for _any_ country listed in [this directory](https://github.com/faasm/covid-sim/tree/faasm/data/admin_units).

File: run/all_native.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+# Experiment information
+IMAGE_NAME = "experiment-covid-native"
+PROJ_ROOT = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+BASE_DIR = "{}/../..".format(PROJ_ROOT)
+with open(os.path.join(BASE_DIR, "VERSION")) as version_file:
+    VERSION = version_file.read().strip()
+RESULTS_FILE = "{}/results/covid/covid_native.dat".format(BASE_DIR)
+
+# Benchmark parameters
```
Comment (csegarragonz) [inline run/all_native.py:16]: These are the parameters that I mention could (should) eventually be moved elsewhere. I think, deciding where to host all of these is still TBD.

File: run/single_native.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        /A:/code/experiment-covid/third-party/covid-sim/data/admin_units/Virgin_Islands_US_admin.txt \
+        /PP:/code/experiment-covid/third-party/covid-sim/data/param_files/preUK_R0=2.0.txt \
+        /P:/code/experiment-covid/third-party/covid-sim/data/param_files/p_NoInt.txt \
+        /O:/tmp/Virgin_Islands_US_NoInt_R0=3.0 \
+        /D:/tmp/wpop_us_terr.txt \
+        /M:/tmp/Virgin_Islands_US_pop_density.bin \
+        /S:/tmp/Network_Virgin_Islands_US_T1_R3.0.bin \
+        /R:1.5 98798150 729101 17389101 4797132
+"""
+
```
Comment (csegarragonz) [inline run/single_native.py]: These whitelines are generated by the python's `black`.

File: run/single_native.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        /c:1 \
+        /A:/code/experiment-covid/third-party/covid-sim/data/admin_units/Virgin_Islands_US_admin.txt \
+        /PP:/code/experiment-covid/third-party/covid-sim/data/param_files/preUK_R0=2.0.txt \
+        /P:/code/experiment-covid/third-party/covid-sim/data/param_files/p_NoInt.txt \
+        /O:/tmp/Virgin_Islands_US_NoInt_R0=3.0 \
+        /D:/tmp/wpop_us_terr.txt \
+        /M:/tmp/Virgin_Islands_US_pop_density.bin \
+        /S:/tmp/Network_Virgin_Islands_US_T1_R3.0.bin \
+        /R:1.5 98798150 729101 17389101 4797132
+"""
```
Comment (Shillaker) [inline run/single_native.py]: Is this documentation or a comment relating to some code? If it's docs you could put it in the README

File: docker/experiment-covid-native.dockerfile [hunk 6]
```diff
@@ -0,0 +1,23 @@
+ARG EXPERIMENTS_VERSION
```
Comment (Shillaker) [inline docker/experiment-covid-native.dockerfile:1]: Do we need two separate Dockerfiles and images for each experiment or can we have one? It's starting to seem like overkill to me. Could we merge the two experiment base containers? The responsibilities for the two are different, but I'm not sure it's worth the extra admin and more difficult development to maintain two images.

We could just have a single `Dockerfile` at the top level of each experiment repo and build a single image.

Afaict the responsibilities of this single container would be the following, which I don't think cause any conflicts:

- Build and run the native version (i.e. check out code, build with standard toolchain, execute)
- Build and upload the wasm version (i.e. check out code, build with Faasm toolchain, upload)

**Review by Shillaker — CHANGES_REQUESTED** [review]
Nice, looks good, just one potential change request, do we still think having two images for each experiment is a good idea? I'm starting to question it. See comment in the review.

File: docker/experiment-covid-native.dockerfile (hunk 6 above)
Comment (csegarragonz) [reply docker/experiment-covid-native.dockerfile:1]: Will do this in faasm/experiment-base#7. In the meantime we can take [this Dockerfile](https://github.com/faasm/experiment-lammps/blob/master/Dockerfile) as reference.

**Review by Shillaker — APPROVED** [review]

**Pull request merged**